File: python/year_2015/day_14.py
"""
https://adventofcode.com/2015/day/14
"""


from pathlib import Path
from itertools import accumulate, chain, cycle, islice, repeat, starmap
import re
from statistics import mode

from iteration_utilities import argmax

R = re.compile(
    (
        r"(?:\w+) can fly (\d+) km/s for (\d+) "
        r"seconds, but then must rest for (\d+) seconds."
    )
)


def process(text):
    def stats(line):
        return tuple(map(int, R.match(line).groups()))

    return tuple(map(stats, text.splitlines()))


def part_1(text, total=2503):
    def distance(speed, flying, resting):
        cycles, extra = divmod(total, flying + resting)
        result = speed * (flying * cycles + min(flying, extra))
        return result

    return max(starmap(distance, process(text)))


def part_2(text, total=2503):
    def distances(speed, flying, resting):
        iterable = chain(repeat(speed, flying), repeat(0, resting))
        yield from accumulate(cycle(iterable))

    iterable = zip(*starmap(distances, process(text)))
    counts = tuple(map(argmax, islice(iterable, total)))

    return counts.count(mode(counts))


def _read_input() -> str:
    return (Path(__file__).with_name("test_data") / "day_14.txt").read_text()


EXAMPLE = """\
Comet can fly 14 km/s for 10 seconds, but then must rest for 127 seconds.
Dancer can fly 16 km/s for 11 seconds, but then must rest for 162 seconds.
"""


def test_part_1_input():
    assert part_1(_read_input()) == 2696

def test_part_1_example():
    assert part_1(EXAMPLE, 1000) == 1120

def test_part_2_input():
    assert part_2(_read_input()) == 1084

def test_part_2_example():
    assert part_2(EXAMPLE, 1000) == 688

File: rust/year_2015/src/day_14.rs
use itertools::Itertools as _;
use num::Integer as _;

pub fn part_1(input: &str) -> u32 {
    max_distance_1(input, 2503)
}

fn max_distance_1(input: &str, n: u32) -> u32 {
    parse_input(input)
        .map(|(speed, fly, rest)| {
            let cycle = fly + rest;
            let (q, r) = n.div_rem(&cycle);
            let t = q * fly + r.min(fly);
            t * speed
        })
        .max()
        .unwrap()
}

pub fn part_2(input: &str) -> u32 {
    max_distance_2(input, 2503)
}

fn max_distance_2(input: &str, n: u32) -> u32 {
    let reindeers: Vec<_> = parse_input(input).collect();
    (1..=n)
        .fold(vec![0; reindeers.len()], |mut points, t| {
            points
                .iter_mut()
                .zip(&reindeers)
                .max_set_by_key(|&(_, &(speed, fly, res))| {
                    let cycle = fly + res;
                    let (q, r) = t.div_rem(&cycle);
                    let t = q * fly + r.min(fly);
                    t * speed
                })
                .into_iter()
                .for_each(|(points, _)| *points += 1);
            points
        })
        .into_iter()
        .max()
        .unwrap()
}

fn parse_input(input: &str) -> impl Iterator<Item = (u32, u32, u32)> {
    input.lines().map(|line| {
        let (_name, rest) = line.split_once(" can fly ").unwrap();
        let (speed, rest) = rest.split_once(" km/s for ").unwrap();
        let (fly, rest) = rest
            .split_once(" seconds, but then must rest for ")
            .unwrap();
        let rest = rest.strip_suffix(" seconds.").unwrap();
        (
            speed.parse().unwrap(),
            fly.parse().unwrap(),
            rest.parse().unwrap(),
        )
    })
}

#[cfg(test)]
mod test {
    use test_case::test_case;

    const INPUT: &str = include_str!("../test_data/day_14.txt");

    const EXAMPLE: &str =
        "Comet can fly 14 km/s for 10 seconds, but then must rest for 127 seconds.
Dancer can fly 16 km/s for 11 seconds, but then must rest for 162 seconds.";

    #[test_case(EXAMPLE => 1120)]
    fn max_distance_1(input: &str) -> u32 {
        super::max_distance_1(input, 1000)
    }

    #[test_case(INPUT => 2696)]
    fn part_1(input: &str) -> u32 {
        super::part_1(input)
    }

    #[test_case(EXAMPLE => 689)]
    fn max_distance_2(input: &str) -> u32 {
        super::max_distance_2(input, 1000)
    }

    #[test_case(INPUT => 1084)]
    fn part_2(input: &str) -> u32 {
        super::part_2(input)
    }
}
